Service CRUD (2)

qangdev opened the pull request.

Qsnake138 committed:
Services CRUD

qangdev committed:
Updated: Submit add/edit/delete service using normal form submit (not AJAX) + services routes. Price data type to dict

qangdev committed:
Updated: main.py changed images to only have file name

mohanadkaleia commented on app/models/services.py:
```diff
@@ -31,13 +28,23 @@ def insert(shop_id="", name="", description="", images=None):
+        "description": description,
+        "price": price,
+        "images": images,
+    }
```
We have few issues with this method:
1. the shop_id is not used?
2. we are not checking or validating the required fields, for example the `id` field is required here
3. in order for this method to work in its current implementation, we would need to always provide all data regardless if they are updated or not, I prefer to only pass the data we need to update, for example if I need to update only the name, then no need to pass the images. To fix it, we will need to build the doc based on the provided args

for example: 
```
doc = {}
if name:
     doc['name'] = name

if description:
    doc['description] = description

....
```

qangdev replied:
1. the shop_id is not used? >>> Nice catch. I'll fix that
2. The id is always included in the request because it's a hidden field on the edit form, so I think only need to remove default value of id in update function. What do you think?
3. This is a good idea, I'll do it.

mohanadkaleia replied:
sounds good! thank you!

mohanadkaleia commented on app/models/services.py:
```diff
@@ -48,10 +55,6 @@ def delete(id):
     results = services.all()
-
-    if not results:
-        raise ErrNotFound("oops.. no service found in the db")
```
nice!

qangdev replied:
I know :))

mohanadkaleia commented on app/views.py:
```diff
@@ -78,6 +78,106 @@ def admin():
+        # get current set of images
+        images = set(service["images"])
+        # add updated images if any
+        images |= set(updated_images)
```
hmmmm first time seeing this `|=` can you explain what does it do?

qangdev replied:
the `|=` is union operation of set data type. The above meaning that I take current list of images of the service union with the new uploaded images. I'm using set to avoid duplicated image names

mohanadkaleia replied:
nice! I did not know that! thank you!

mohanadkaleia commented on app/views.py:
```diff
@@ -78,6 +78,106 @@ def admin():
+def admin_services_images_remove(id):
+    try:
+        # remove images from disk if any
+        rm_image = request.form.get("image")
```
hmmm is this reading and deleting images from the form? won't be easier to read the images values from the service after we get it by id?

mohanadkaleia commented on app/views.py:
```diff
@@ -78,6 +78,106 @@ def admin():
+
+
+@app.route("/admin/services/<id>/delete", methods=["POST"])
+def admin_services_deleting(id):
```
small one, just need to update the method name `deleting` -> `delete`

qangdev replied:
I'll update it

qangdev committed:
method name deleting -> delete
in update service, build doc to list field need to be updated only
some minor changes

mohanadkaleia commented:
Thank you, I checked the PR, and also I checked the interface, here is a quick feedback on it:
I still can't add more than one image per service from admin
I still can't find any delete button for images, or services

qangdev committed:
Merge branch 'master' of https://github.com/mohanadkaleia/gears

# Conflicts:
#	.gitignore
#	cmd/main.py
